lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -181,6 +181,229 @@
         }
     });
 
+    platform.registerChakraLibraryFunction("DefaultMergeSort", function(array, length) {
+        const compareArray = [];
+        for (let index = 0; index < length; ++index) {
+            __chakraLibrary.arrayCreateDataPropertyOrThrow(compareArray, index, "" + array[index]);
+        }
+
+        const buffer = [];
+        const stringBuffer = [];
+        let i = 1, j = 0, k = 0, bucketSize = 4, lastSize = 2, position = 0;
+
+        // handle pairs first
+        while (i < length) {
+            if (compareArray[i] < compareArray[i - 1]) {
+                const cache = array[i];
+                array[i] = array[i - 1];
+                array[i - 1] = cache;
+
+                const stringCache = compareArray[i];
+                compareArray[i] = compareArray[i - 1];
+                compareArray[i - 1] = stringCache;
+            }
+            i += 2;
+        }
+        const doubleLength = length * 2;
+
+        while (bucketSize < doubleLength) {
+            while (position < length) {
+                const left = position;
+                let right = position + bucketSize;
+                const mid = left + lastSize;
+
+                // perform a merge but only if it's necessary
+                if (mid < length && compareArray[mid] < compareArray[mid - 1]) {
```

Comment:
Could compareArray[mid] be out of bounds here? Like your "surprise" test case, but with 5 elements instead of 7?

Reply:
Never mind, apparently I didn't read the first part of this line.

---
In reply to: [220293493](https://github.com/Microsoft/ChakraCore/pull/5724#discussion_r220293493) [](ancestors = 220293493)